util/src/main/java/com/redhat/lightblue/util/MemoryMonitor.java

```diff
@@ -0,0 +1,85 @@
+package com.redhat.lightblue.util;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Utility class for tracking memory usage for given type (assuming it's an immutable value).
+ * You can set thresholds and have it execute callbacks when they are exceeded. See {@link MemoryMonitorTest} for usage.
+ *
+ * @author mpatercz
+ *
+ * @param <T>
+ */
+public class MemoryMonitor<T> {
+
+    @FunctionalInterface
+    public static interface SizeCalculator<T> {
+        public int size(T obj);
+    }
+
+    @FunctionalInterface
+    public static interface ThresholdExceededCallback<T> {
+        public void fire(int currentSizeB, int thresholdB, T obj) throws RuntimeException;
+    }
+
+    public static class ThresholdMonitor<T> {
+
+        final int thresholdB;
+
+        boolean fired = false;
+
+        final ThresholdExceededCallback<T> event;
+
+        public ThresholdMonitor(int thresholdB, ThresholdExceededCallback<T> event) {
+            super();
+            this.thresholdB = thresholdB;
+            this.event = event;
+        }
+
+    }
+
+    private int dataSizeB = 0;
+
+    private SizeCalculator<T> sizeCalculator;
+
+    private List<ThresholdMonitor<T>> monitors = new ArrayList<>();
+
+    public void registerMonitor(ThresholdMonitor<T> m) {
+        if (m.thresholdB > 0) {
+            this.monitors.add(m);
+        }
+    }
+
+    private void checkThresholdMonitors(T obj) throws RuntimeException {
+        for (ThresholdMonitor<T> m: monitors) {
+            if (dataSizeB > m.thresholdB && !m.fired) {
+                m.fired = true;
+                m.event.fire(dataSizeB, m.thresholdB, obj);
```

Comment:
May be worth capturing a TODO or something here that the first monitor to throw an exception will stop all other callbacks from firing. As a general purpose class, this is probably surprising, though we know with its current usages it may not be problematic. So, I suggest at least a TODO and/or open an issue to remind ourselves that this should probably catch any thrown exceptions and aggregate them.